extensions/Lily/MoreEvents.js

```diff
@@ -0,0 +1,424 @@
+(function (Scratch) {
+  'use strict';
+
+  const vm = Scratch.vm;
+  const runtime = vm.runtime;
+
+  const stopIcon = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAMAAAAp4XiDAAAAQlBMVEUAAAC/UFC8Q0OzTU24SEi4SEi3SEi4R0e4SEi4SEi4SEi4SEi7SUm8SUnMTk7MT0/OT0/PT0/gVVXiVVXsWVn///+CoOd2AAAAC3RSTlMAEBMUu7zLz9D8/dIXnJwAAAABYktHRBXl2PmjAAAAxklEQVRIx+3WwRKDIBAD0JWqVEOtWv7/W3twOqKwELzW3N9wYhORMMYiztgZUZMUAKxqmh5Kno/MG256nzI59Z2mB+BWH+XzUt5RhWoyQjFZkTQFkTBFERlCnAwlDoYUgaHFblpaeL86AK0MvNjMIABmT2cGIAAWniw3ucm/k9ovduEjXzgXtUfJmtrTt9VZzYH9FSB/xvfKZMsiLFmuko61zBTfucjL9RpXf6nEU2MhPxXS86J+kORmjz6V6seViOnG8oT7ApMcjsYZwhXCAAAAAElFTkSuQmCC';
+
+  // Source:
+  // https://github.com/TurboWarp/scratch-vm/blob/develop/src/io/keyboard.js
+  // https://github.com/TurboWarp/scratch-blocks/blob/develop/blocks_vertical/event.js
+  const validKeyboardInputs = [
+    // Special Inputs
+    { text: 'space', value: 'space' },
+    { text: 'up arrow', value: 'up arrow' },
+    { text: 'down arrow', value: 'down arrow' },
+    { text: 'right arrow', value: 'right arrow' },
+    { text: 'left arrow', value: 'left arrow' },
+    { text: 'enter', value: 'enter' },
+    // TW: Extra Special Inputs
+    { text: 'backspace', value: 'backspace' },
+    { text: 'delete', value: 'delete' },
+    { text: 'shift', value: 'shift' },
+    { text: 'caps lock', value: 'caps lock' },
+    { text: 'scroll lock', value: 'scroll lock' },
+    { text: 'control', value: 'control' },
+    { text: 'escape', value: 'escape' },
+    { text: 'insert', value: 'insert' },
+    { text: 'home', value: 'home' },
+    { text: 'end', value: 'end' },
+    { text: 'page up', value: 'page up' },
+    { text: 'page down', value: 'page down' },
+    // Letter Keyboard Inputs
+    { text: 'a', value: 'a' },
+    { text: 'b', value: 'b' },
+    { text: 'c', value: 'c' },
+    { text: 'd', value: 'd' },
+    { text: 'e', value: 'e' },
+    { text: 'f', value: 'f' },
+    { text: 'g', value: 'g' },
+    { text: 'h', value: 'h' },
+    { text: 'i', value: 'i' },
+    { text: 'j', value: 'j' },
+    { text: 'k', value: 'k' },
+    { text: 'l', value: 'l' },
+    { text: 'm', value: 'm' },
+    { text: 'n', value: 'n' },
+    { text: 'o', value: 'o' },
+    { text: 'p', value: 'p' },
+    { text: 'q', value: 'q' },
+    { text: 'r', value: 'r' },
+    { text: 's', value: 's' },
+    { text: 't', value: 't' },
+    { text: 'u', value: 'u' },
+    { text: 'v', value: 'v' },
+    { text: 'w', value: 'w' },
+    { text: 'x', value: 'x' },
+    { text: 'y', value: 'y' },
+    { text: 'z', value: 'z' },
+    // Number Keyboard Inputs
+    { text: '0', value: '0' },
+    { text: '1', value: '1' },
+    { text: '2', value: '2' },
+    { text: '3', value: '3' },
+    { text: '4', value: '4' },
+    { text: '5', value: '5' },
+    { text: '6', value: '6' },
+    { text: '7', value: '7' },
+    { text: '8', value: '8' },
+    { text: '9', value: '9' },
+  ];
+
+  setInterval(() => {

```

Comment:
setInterval like this significantly impacts battery life. use vm.on('BEFORE_STEP', () => {}) instead (from memory, it's something like that, animated text uses it)

Reply:
its BEFORE_EXECUTE, the stop sign code below already uses it